File: basic_dsa_nineteen.java
/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */
import java.util.*;

public class basic_dsa_nineteen {
    
    // Problem 1: Check if a Number is Abundant Number
    // An abundant number is a number for which the sum of its proper divisors is greater than the number itself
    // Example: 12 has proper divisors 1, 2, 3, 4, 6, sum = 16 > 12, so 12 is abundant
    
    /**
     * Calculate sum of proper divisors of a number
     * Time Complexity: O(√n)
     * Space Complexity: O(1)
     */
    public static int sumOfProperDivisors(int n) {
        if (n <= 1) {
            return 0;
        }
        
        int sumDivisors = 1; // 1 is always a divisor
        
        // Check divisors from 2 to √n
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0) {
                sumDivisors += i;
                // Add the corresponding divisor (avoid adding square twice)
                if (i != n / i) {
                    sumDivisors += n / i;
                }
            }
        }
        
        return sumDivisors;
    }
    
    /**
     * Check if a number is abundant number
     * Time Complexity: O(√n)
     * Space Complexity: O(1)
     */
    public static boolean isAbundantNumber(int n) {
        if (n <= 0) {
            return false;
        }
        return sumOfProperDivisors(n) > n;
    }
    
    /**
     * Get all proper divisors of a number
     * Time Complexity: O(√n)
     * Space Complexity: O(√n)
     */
    public static ArrayList<Integer> getProperDivisors(int n) {
        ArrayList<Integer> divisors = new ArrayList<>();
        if (n <= 1) {
            return divisors;
        }
        
        divisors.add(1);
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0) {
                divisors.add(i);
                if (i != n / i) {
                    divisors.add(n / i);
                }
            }
        }
        
        Collections.sort(divisors);
        return divisors;
    }
    
    // Problem 2: Find All Prime Numbers in a Range
    // Find all prime numbers between two given numbers using Sieve of Eratosthenes
    
    /**
     * Find all prime numbers up to n using Sieve of Eratosthenes
     * Time Complexity: O(n log log n)
     * Space Complexity: O(n)
     */
    public static ArrayList<Integer> sieveOfEratosthenes(int n) {
        if (n < 2) {
            return new ArrayList<>();
        }
        
        // Create boolean array, initially all true
        boolean[] isPrime = new boolean[n + 1];
        Arrays.fill(isPrime, true);
        isPrime[0] = isPrime[1] = false;
        
        // Mark multiples of primes as false
        for (int i = 2; i * i <= n; i++) {
            if (isPrime[i]) {
                // Mark all multiples of i as non-prime
                for (int j = i * i; j <= n; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        
        // Collect all prime numbers
        ArrayList<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= n; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }
        
        return primes;
    }
    
    /**
     * Find all prime numbers in a given range [start, end]
     * Time Complexity: O(end log log end) for sieve + O(end - start) for filtering
     * Space Complexity: O(end)
     */
    public static ArrayList<Integer> primesInRange(int start, int end) {
        if (end < 2 || start > end) {
            return new ArrayList<>();
        }
        
        // Adjust start to be at least 2
        start = Math.max(2, start);
        
        // Use sieve to find all primes up to end
        ArrayList<Integer> allPrimes = sieveOfEratosthenes(end);
        
        // Filter primes in the range [start, end]
        ArrayList<Integer> primesInRangeList = new ArrayList<>();
        for (int prime : allPrimes) {
            if (prime >= start) {
                primesInRangeList.add(prime);
            }
        }
        
        return primesInRangeList;
    }
    
    /**
     * Simple prime check for a single number
     * Time Complexity: O(√n)
     * Space Complexity: O(1)
     */
    public static boolean isPrimeSimple(int n) {
        if (n < 2) {
            return false;
        }
        if (n == 2) {
            return true;
        }
        if (n % 2 == 0) {
            return false;
        }
        
        for (int i = 3; i * i <= n; i += 2) {
            if (n % i == 0) {
                return false;
            }
        }
        
        return true;
    }
    
    /**
     * Count prime numbers in a given range
     * Time Complexity: O(end log log end)
     * Space Complexity: O(end)
     */
    public static int countPrimesInRange(int start, int end) {
        ArrayList<Integer> primes = primesInRange(start, end);
        return primes.size();
    }
    
    // Test cases
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem 1: Check if a Number is Abundant Number");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Abundant number 12
        int num1 = 12;
        boolean result1 = isAbundantNumber(num1);
        ArrayList<Integer> divisors1 = getProperDivisors(num1);
        int sum1 = divisors1.stream().mapToInt(Integer::intValue).sum();
        System.out.println("Number: " + num1);
        System.out.println("Proper divisors: " + divisors1);
        System.out.println("Sum of proper divisors: " + sum1);
        System.out.println("Is abundant: " + result1);
        
        // Test Case 2 - Not abundant number
        int num2 = 10;
        boolean result2 = isAbundantNumber(num2);
        ArrayList<Integer> divisors2 = getProperDivisors(num2);
        int sum2 = divisors2.stream().mapToInt(Integer::intValue).sum();
        System.out.println("\nNumber: " + num2);
        System.out.println("Proper divisors: " + divisors2);
        System.out.println("Sum of proper divisors: " + sum2);
        System.out.println("Is abundant: " + result2);
        
        // Test Case 3 - Abundant number 18
        int num3 = 18;
        boolean result3 = isAbundantNumber(num3);
        ArrayList<Integer> divisors3 = getProperDivisors(num3);
        int sum3 = divisors3.stream().mapToInt(Integer::intValue).sum();
        System.out.println("\nNumber: " + num3);
        System.out.println("Proper divisors: " + divisors3);
        System.out.println("Sum of proper divisors: " + sum3);
        System.out.println("Is abundant: " + result3);
        
        // Test Case 4 - Perfect number (not abundant)
        int num4 = 6;
        boolean result4 = isAbundantNumber(num4);
        ArrayList<Integer> divisors4 = getProperDivisors(num4);
        int sum4 = divisors4.stream().mapToInt(Integer::intValue).sum();
        System.out.println("\nNumber: " + num4);
        System.out.println("Proper divisors: " + divisors4);
        System.out.println("Sum of proper divisors: " + sum4);
        System.out.println("Is abundant: " + result4);
        
        // Test Case 5 - Deficient number (not abundant)
        int num5 = 8;
        boolean result5 = isAbundantNumber(num5);
        ArrayList<Integer> divisors5 = getProperDivisors(num5);
        int sum5 = divisors5.stream().mapToInt(Integer::intValue).sum();
        System.out.println("\nNumber: " + num5);
        System.out.println("Proper divisors: " + divisors5);
        System.out.println("Sum of proper divisors: " + sum5);
        System.out.println("Is abundant: " + result5);
        
        System.out.println("\n" + "=".repeat(60));
        System.out.println("Problem 2: Find All Prime Numbers in a Range");
        System.out.println("=".repeat(60));
        
        // Test Case 1 - Primes up to 30
        int n1 = 30;
        ArrayList<Integer> primes1 = sieveOfEratosthenes(n1);
        System.out.println("Prime numbers up to " + n1 + ":");
        System.out.println(primes1);
        System.out.println("Count: " + primes1.size());
        
        // Test Case 2 - Primes in range [10, 50]
        int start2 = 10;
        int end2 = 50;
        ArrayList<Integer> primes2 = primesInRange(start2, end2);
        System.out.println("\nPrime numbers in range [" + start2 + ", " + end2 + "]:");
        System.out.println(primes2);
        System.out.println("Count: " + primes2.size());
        
        // Test Case 3 - Primes in range [1, 20]
        int start3 = 1;
        int end3 = 20;
        ArrayList<Integer> primes3 = primesInRange(start3, end3);
        System.out.println("\nPrime numbers in range [" + start3 + ", " + end3 + "]:");
        System.out.println(primes3);
        System.out.println("Count: " + primes3.size());
        
        // Test Case 4 - Count primes in range
        int start4 = 2;
        int end4 = 100;
        int count4 = countPrimesInRange(start4, end4);
        System.out.println("\nNumber of primes in range [" + start4 + ", " + end4 + "]: " + count4);
        
        // Test Case 5 - Primes up to 100
        int n5 = 100;
        ArrayList<Integer> primes5 = sieveOfEratosthenes(n5);
        System.out.println("\nPrime numbers up to " + n5 + ":");
        System.out.println(primes5);
        System.out.println("Count: " + primes5.size());
    }
}

File: basic_dsa_nineteen.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Check if a Number is Abundant Number
# An abundant number is a number for which the sum of its proper divisors is greater than the number itself
# Example: 12 has proper divisors 1, 2, 3, 4, 6, sum = 16 > 12, so 12 is abundant

def sum_of_proper_divisors(n):
    """
    Calculate sum of proper divisors of a number
    Time Complexity: O(√n)
    Space Complexity: O(1)
    """
    if n <= 1:
        return 0
    
    sum_divisors = 1  # 1 is always a divisor
    
    # Check divisors from 2 to √n
    i = 2
    while i * i <= n:
        if n % i == 0:
            sum_divisors += i
            # Add the corresponding divisor (avoid adding square twice)
            if i != n // i:
                sum_divisors += n // i
        i += 1
    
    return sum_divisors


def is_abundant_number(n):
    """
    Check if a number is abundant number
    Time Complexity: O(√n)
    Space Complexity: O(1)
    """
    if n <= 0:
        return False
    return sum_of_proper_divisors(n) > n


def get_proper_divisors(n):
    """
    Get all proper divisors of a number
    Time Complexity: O(√n)
    Space Complexity: O(√n)
    """
    if n <= 1:
        return []
    
    divisors = [1]
    i = 2
    
    while i * i <= n:
        if n % i == 0:
            divisors.append(i)
            if i != n // i:
                divisors.append(n // i)
        i += 1
    
    divisors.sort()
    return divisors


# Problem 2: Find All Prime Numbers in a Range
# Find all prime numbers between two given numbers using Sieve of Eratosthenes

def sieve_of_eratosthenes(n):
    """
    Find all prime numbers up to n using Sieve of Eratosthenes
    Time Complexity: O(n log log n)
    Space Complexity: O(n)
    """
    if n < 2:
        return []
    
    # Create boolean array, initially all True
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    
    # Mark multiples of primes as False
    i = 2
    while i * i <= n:
        if is_prime[i]:
            # Mark all multiples of i as non-prime
            for j in range(i * i, n + 1, i):
                is_prime[j] = False
        i += 1
    
    # Collect all prime numbers
    primes = [num for num in range(2, n + 1) if is_prime[num]]
    return primes


def primes_in_range(start, end):
    """
    Find all prime numbers in a given range [start, end]
    Time Complexity: O(end log log end) for sieve + O(end - start) for filtering
    Space Complexity: O(end)
    """
    if end < 2 or start > end:
        return []
    
    # Adjust start to be at least 2
    start = max(2, start)
    
    # Use sieve to find all primes up to end
    all_primes = sieve_of_eratosthenes(end)
    
    # Filter primes in the range [start, end]
    primes_in_range_list = [p for p in all_primes if p >= start]
    return primes_in_range_list


def is_prime_simple(n):
    """
    Simple prime check for a single number
    Time Complexity: O(√n)
    Space Complexity: O(1)
    """
    if n < 2:
        return False
    if n == 2:
        return True
    if n % 2 == 0:
        return False
    
    i = 3
    while i * i <= n:
        if n % i == 0:
            return False
        i += 2
    
    return True


def count_primes_in_range(start, end):
    """
    Count prime numbers in a given range
    Time Complexity: O(end log log end)
    Space Complexity: O(end)
    """
    primes = primes_in_range(start, end)
    return len(primes)


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Check if a Number is Abundant Number")
    print("=" * 60)
    
    # Test Case 1 - Abundant number 12
    num1 = 12
    result1 = is_abundant_number(num1)
    divisors1 = get_proper_divisors(num1)
    sum1 = sum(divisors1)
    print(f"Number: {num1}")
    print(f"Proper divisors: {divisors1}")
    print(f"Sum of proper divisors: {sum1}")
    print(f"Is abundant: {result1}")
    
    # Test Case 2 - Not abundant number
    num2 = 10
    result2 = is_abundant_number(num2)
    divisors2 = get_proper_divisors(num2)
    sum2 = sum(divisors2)
    print(f"\nNumber: {num2}")
    print(f"Proper divisors: {divisors2}")
    print(f"Sum of proper divisors: {sum2}")
    print(f"Is abundant: {result2}")
    
    # Test Case 3 - Abundant number 18
    num3 = 18
    result3 = is_abundant_number(num3)
    divisors3 = get_proper_divisors(num3)
    sum3 = sum(divisors3)
    print(f"\nNumber: {num3}")
    print(f"Proper divisors: {divisors3}")
    print(f"Sum of proper divisors: {sum3}")
    print(f"Is abundant: {result3}")
    
    # Test Case 4 - Perfect number (not abundant)
    num4 = 6
    result4 = is_abundant_number(num4)
    divisors4 = get_proper_divisors(num4)
    sum4 = sum(divisors4)
    print(f"\nNumber: {num4}")
    print(f"Proper divisors: {divisors4}")
    print(f"Sum of proper divisors: {sum4}")
    print(f"Is abundant: {result4}")
    
    # Test Case 5 - Deficient number (not abundant)
    num5 = 8
    result5 = is_abundant_number(num5)
    divisors5 = get_proper_divisors(num5)
    sum5 = sum(divisors5)
    print(f"\nNumber: {num5}")
    print(f"Proper divisors: {divisors5}")
    print(f"Sum of proper divisors: {sum5}")
    print(f"Is abundant: {result5}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Find All Prime Numbers in a Range")
    print("=" * 60)
    
    # Test Case 1 - Primes up to 30
    n1 = 30
    primes1 = sieve_of_eratosthenes(n1)
    print(f"Prime numbers up to {n1}:")
    print(primes1)
    print(f"Count: {len(primes1)}")
    
    # Test Case 2 - Primes in range [10, 50]
    start2 = 10
    end2 = 50
    primes2 = primes_in_range(start2, end2)
    print(f"\nPrime numbers in range [{start2}, {end2}]:")
    print(primes2)
    print(f"Count: {len(primes2)}")
    
    # Test Case 3 - Primes in range [1, 20]
    start3 = 1
    end3 = 20
    primes3 = primes_in_range(start3, end3)
    print(f"\nPrime numbers in range [{start3}, {end3}]:")
    print(primes3)
    print(f"Count: {len(primes3)}")
    
    # Test Case 4 - Count primes in range
    start4 = 2
    end4 = 100
    count4 = count_primes_in_range(start4, end4)
    print(f"\nNumber of primes in range [{start4}, {end4}]: {count4}")
    
    # Test Case 5 - Primes up to 100
    n5 = 100
    primes5 = sieve_of_eratosthenes(n5)
    print(f"\nPrime numbers up to {n5}:")
    print(primes5)
    print(f"Count: {len(primes5)}")
